### ISSUE TITLE
mongodb not using env variable for DATABASE_URL

### ISSUE THREAD

**Issue by janiukjf** [db-migrate/node-db-migrate#629]
## I'm submitting a...

 - [x] Bug report 
 - [ ] Feature request
 - [ ] Question

## Current behavior
We're running mongodb and have a database json that looks like this:
`"defaultEnv": "dev",
  "dev": {
    "driver": "mongodb",
    "DATABASE_URL": {"ENV": "MONGO_URL"}
  },
`
It doesn't seem to matter what ENV variable we set, we always get Error: database must be defined in database.json. If we manually set DATABASE_URL in our bash env, it works fine, but it doesn't transform using MONGO_URL or really any env variable we set up. It seems like it's just ignoring the environment transformation entirely.

Here's the full error:
```
[ERROR] Error: database must be defined in database.json
    at Object.exports.connect (/Users/Janiuk/repos/graphql-server/node_modules/db-migrate-mongodb/index.js:496:11)
    at connect (/Users/Janiuk/repos/graphql-server/node_modules/db-migrate/lib/driver/index.js:93:12)
    at Object.exports.connect (/Users/Janiuk/repos/graphql-server/node_modules/db-migrate/lib/driver/index.js:134:5)
    at Object.exports.connect (/Users/Janiuk/repos/graphql-server/node_modules/db-migrate/connect.js:15:10)
    at /Users/Janiuk/repos/graphql-server/node_modules/db-migrate/lib/commands/up.js:15:13
    at tryCatcher (/Users/Janiuk/repos/graphql-server/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/Janiuk/repos/graphql-server/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/Users/Janiuk/repos/graphql-server/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromiseCtx (/Users/Janiuk/repos/graphql-server/node_modules/bluebird/js/release/promise.js:606:10)
    at Async._drainQueue (/Users/Janiuk/repos/graphql-server/node_modules/bluebird/js/release/async.js:138:12)
    at Async._drainQueues (/Users/Janiuk/repos/graphql-server/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/Janiuk/repos/graphql-server/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
```

## Expected behavior
DATABASE_URL should be properly populated using the ENV setting.

## Minimal reproduction of the problem with instructions
Create a project with mongodb and db-migrate. Create a database.json looking something a-like-so:

```
{
  "defaultEnv": "dev",
  "dev": {
    "driver": "mongodb",
    "DATABASE_URL": {"ENV": "MONGO_URL"}
  },
  "staging": {
    "driver": "mongodb",
    "DATABASE_URL": {"ENV": "STAGING_DATABASE_URL"}
  },
  "uat": {
    "driver": "mongodb",
    "DATABASE_URL": {"ENV": "UAT_DATABASE_URL"}
  },
  "demo": {
    "driver": "mongodb",
    "DATABASE_URL": {"ENV": "DEMO_DATABASE_URL"}
  },
  "prod": {
    "driver": "mongodb",
    "DATABASE_URL": {"ENV": "PROD_DATABASE_URL"}
  }
}
```
Run migrations using `npx db-migrate up`. Watch the error listed above occur.

## What is the motivation / use case for changing the behavior?
We'd love to see A: the environment variable properly populated when specifying it above and B: Not have to duplicate our environment variables with both MONGO_URL and DATABASE_URL with the same information.

## Environment

<pre><code>
db-migrate version: 0.11.6
plugins with versions:
db-migrate-mongodb: 1.5.0
db-migrate driver with versions: mongodb 3.4.6

Additional information:
- Node version: 10.15.1
- Platform:  Mac

Others:
</code></pre>

**Comment by janiukjf** [comment]
This is kind of a mix of bug-report / feature-request / question. We might just be doing it wrong.

**Comment by gmahe** [comment]
I don't think it is how you do it, it should be that way:
```
  {
  "defaultEnv": "dev",
  "dev": { "ENV": "MONGO_URL" },
  "staging": { "ENV": "STAGING_DATABASE_URL" },
  "uat": { "ENV": "UAT_DATABASE_URL" },
  "demo": { "ENV": "DEMO_DATABASE_URL" },
  "prod": { "ENV": "PROD_DATABASE_URL" }
}
```
with `export MONGO_URL="mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin"`

**Comment by janiukjf** [comment]
Yeah, I thought that too. I did try this, and it didn't work either. It unfortunately didn't work with any environment variable name I tried. I'm wondering if it has to do with how the mongo driver is configured in the docs, which splits up the parts of the connection string instead of the standard MONGO_URL env variable.

**Comment by gmahe** [comment]
Maybe your environment variables were inside your container and you were running the command from outside or vice-versa

**Comment by janiukjf** [comment]
In the case of me testing this, I wasn't even doing it from in a container. I was running it locally with local MONGO_URL variables set. The only way I was able to get this to work was with DATABASE_URL set, which just bypasses this config entirely. So I'm not sure what's wrong in my setup.

**Comment by gmahe** [comment]
>DATABASE_URL bypassing this config? 

What do you mean?

```export MONGO_URL="mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin"```
this didn't work?

**Comment by janiukjf** [comment]
Sorry for the confusion here. MONGO_URL is definitely set in the environment, as is required to get mongo working anywhere. What I'm saying is the database.json file set like below doesn't read it in.

`"dev": {
    "driver": "mongodb",
    "DATABASE_URL": {"ENV": "MONGO_URL"}
  },`

In regards to the "bypassing this config" statement, in the configuration docs, it states that if you set the environment variable `DATABASE_URL`, it skips the database.json file and just uses the DATABASE_URL environment variable. That's the only way we've been able to get things working with db-migrate both locally and in our containers.

So to clarify, I have both MONGO_URL and DATABASE_URL set, and both set to the same value. I was unable to get it working with the config/database.json as structured above and with no DATABASE_URL manually set in my environment. It seems to ignore whatever I've put in for ENV in the database.json.

Again...apologies for confusion.

**Comment by gmahe** [comment]
So basically this is how it works:
```json
{
  "dev": {} // or whatever value in it
}
```
or
```
{
  "dev": { "ENV": "MONGO_URL" }
}
```
will return `mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin` if you have DATABASE_URL set up via:
```
export DATABASE_URL="mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/admin"
```

If you don't have DATABASE_URL env var setup then it will read you database.json and in your case use `MONGO_URL` which you will have as en env var.


The doc confirms if you have `DATABASE_URL` setup then it will ignore what's in the database.json.

>Alternatively, you can specify a DATABASE_URL environment variable that will be used in place of the configuration file settings. This is helpful for use with Heroku.
>Note: If a database url is specified, the config file is being skipped. You can however also specify rc configs, where you can configure everything you can configure also on the CLI.

**Comment by janiukjf** [comment]
Yes, we are saying the same thing. But, I'm saying also that this:
`{
  "dev": { "ENV": "MONGO_URL" }
}`
is not working for us. We don't know why it won't take in our `MONGO_URL` value here. We just get the error listed above. For some reason it just plain isn't. We spent about a full day messing around with this to no avail. We ended up just setting `DATABASE_URL` manually because we spent too much time spinning our wheels trying to figure out what was wrong. It isn't using the value we have set for `MONGO_URL`. We spent a ton of time in that documentation trying to figure out what wasn't working. We're using `DATABASE_URL` as our workaround, but we'd love to get the config working.

I'll spend some more time on Monday with it to see if I can find out more. It's entirely possible files are in the wrong location. I'll report back if things change.

**Comment by gmahe** [comment]
To me it looks like the URL is wrong but you said it's the same as `DATABASE_URL` and `DATABASE_URL` is working.
So if you've got the `[ERROR] Error: database must be defined in database.json` and that you don't have `DATABASE_URL` setup it means it picks up the config from database.json but the error comes from the url itself, are you sure you didn't missed the last bit after the `/` ?
```
mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/*thisIsTheDB*
```
Are they the exact same value when you echo them?
It doesn't make much sense to me otherwise

I've just tested the code bellow and it works perfectly.
```
{
  "dev": { "ENV": "MONGO_URL" }
}
```
```
export MONGO_URL="mongodb://myDBReader:D1fficultP%40ssw0rd@mongodb0.example.com:27017/*thisIsTheDB*"
```

**Comment by janiukjf** [comment]
I thought about this for a while last night and I have a strong suspicion that the only reason this isn't working is that the database.json is somehow in the wrong place. I'm trying to remember, because it's been a few weeks now, but I don't know that the database.json ever worked for us. I'm going to check our environment setup to make sure files are where they should be for this. If they are, I'll try what you mentioned above and see if that works. I'll get back to you as soon as the US holiday weekend is over.

**Comment by gmahe** [comment]
That could be that too, you need to run the `db-migrate` from the same dir as where your `database.json`is.
Good luck

**Comment by janiukjf** [comment]
OK. So, I experimented this morning, and I got it working properly. I'm certain it was a mix of both the engineer that implemented this initially reading the docs wrong, and then afterwards...me, also reading the docs wrong after viewing what we had and getting confused. There's a section in which user and password are passed in via env variables, and I think that's where the confusion came from. This is the block from the docs:

```
{
  "prod": {
    "driver": "mysql",
    "user": {"ENV": "PRODUCTION_USERNAME"},
    "password": {"ENV": "PRODUCTION_PASSWORD"}
  },
}
```

After our conversation last week, this morning I replaced our database.json config removing the "DATABASE_URL" value that's in the original post above and used basically exactly what you suggested a few days ago that I initially misread as the same thing we had. Once I did that and removed the environment variable of DATABASE_URL, it worked exactly as it's supposed to. So this was a case of poorly reading the docs and then myself poorly following up on it.

Thanks for your patience working through this user error issue. It's much appreciated.
